File: frontend/src/modules/finance.ts
import { catalog } from '../catalog'
import type { Command } from '../types'
import type { Field } from '../types'
import type { Row } from '../types'
import type { Column } from '../types'
import { t } from './shared'
import { date } from './shared'
import { reason } from './shared'
import { finance } from './shared'
import { C } from './shared'
import { project } from './shared'
import { endpoint } from './shared'
import { ledgerStatus } from './shared'
import { all, download, read } from '../api'
import { reconciliationCommand } from './settlement'
export const columns: Record<string, Column[]> = {
  entries: [
    C('project_name', '项目'),
    C('title', '款项'),
    C('kind', '类型'),
    C('amount', '原金额'),
    C('credit_amount', '冲减'),
    C('paid_amount', '已结算'),
    C('balance', '待结算'),
    C('due_date', '最近待付期限'), C('due_amount', '当前到期金额'),
  ],
  payments: [
    C('entry_title', '款项'),
    C('amount', '金额'),
    C('date', '日期'),
    C('reason', '说明'),
    C('method', '结算方式'), C('account', '账户标识'), C('reference', '银行流水号'),
    C('reversal_of', '冲销原记录'),
    C('reconciliation', '对账单ID'),
    { key: 'bank_matched', label: '银行核对', format: r => r.bank_matched ? '已匹配' : '未匹配' },
    { key: 'reversed_by', label: '记录状态', format: ledgerStatus },
  ],
}
export function createLabel(resource: string) {
  return ({ entries: finance() && '登记费用' } as Record<string, string | boolean>)[resource] || ''
}
export async function createCommand(resource: string, projectId?: number): Promise<Command> {
  const c = await catalog(['projects'])
  let fields: Field[] = []
  let path = endpoint(resource)
  if (resource === 'entries') {
    path += 'expense/'
    fields = [
      project(c),
      t('title', '费用名称'),
      t('amount', '金额（元）'),
      date('due_date', '期限'),
    ]
  }
  return {
    title: String(createLabel(resource)),
    path,
    fields,
    initial: projectId ? { project: projectId } : {},
  }
}
export function actionNames(resource: string, r: Row): string[] {
  const a: string[] = []
  if (resource === 'entries') a.push('查看收付流水', '查看到期明细')
  if (resource === 'payments' && r.document) a.push('下载凭证')
  if (resource === 'payments') a.push('查看凭证')
  if (resource === 'payments' && finance()) a.push('补录凭证')
  if (resource === 'entries' && finance()) {
    if (Number(r.balance) !== 0) a.push('生成对账单')
    if (r.purchase && Number(r.balance) > 0) a.push('申请预付款核准')
    if (Number(r.balance) > 0) a.push('登记收付款')
    if (Number(r.balance) < 0) a.push('退款')
    if (r.kind === 'expense' && !r.cancelled) a.push('取消费用')
  }
  if (resource === 'payments' && finance() && !r.reversal_of && !r.reversed_by) a.push('冲销付款')
  return a
}
export async function actionCommand(resource: string, r: Row, name: string): Promise<Command> {
  if (name === '查看到期明细') return { title: name, path: '', readonly: true, initial: { lines: r.payment_schedule || [] }, fields: [{ key: 'lines', label: '未结到期明细', type: 'rows', fields: [t('due_date', '到期日'), t('amount', '待付金额')] }], notice: { type: 'info', text: '月结按合格收货当月月底加账期天数；跨月分别到期。已付款优先抵扣先到期部分，退货冲减原收货批次。未收货部分尚未起算账期，预付款仍需合同核准。' } }
  if (name === '生成对账单' || name === '申请预付款核准') return reconciliationCommand(r, name === '申请预付款核准')
  if (name === '补录凭证') {
    const docs = await all('/business/documents/', { project: r.project, category: 'receipt' })
    return { title: name, path: `/business/payments/${r.id}/attach-evidence/`, fields: [{ key: 'document', label: '凭证（项目附件中上传）', type: 'select', options: docs.map(d => ({ value: d.id, label: d.original_name })) }, reason], notice: { type: 'info', text: '追加凭证并留痕，不覆盖原付款金额和历史凭证。' } }
  }
  if (name === '查看凭证') {
    const detail = await read(`/business/payments/${r.id}/`)
    return { title: name, path: '', readonly: true, initial: { lines: detail.evidence }, fields: [{ key: 'lines', label: '凭证记录', type: 'rows', fields: [t('name', '文件'), t('reason', '说明'), t('actor', '补录人'), t('date', '补录时间')] }], actions: detail.evidence.map((d: Row) => ({ label: `下载 ${d.name}`, run: async () => { const doc = await read(`/business/documents/${d.document}/`); await download(doc.download_url, doc.original_name) } })) }
  }
  if (name === '查看收付流水') return { title: name, path: '', readonly: true, initial: { lines: await all('/business/payments/', { entry: r.id }) }, fields: [{ key: 'lines', label: '收付流水', type: 'rows', fields: [t('id', '流水ID'), t('date', '日期'), t('amount', '金额'), t('method', '方式'), t('account', '账户'), t('reference', '银行流水号'), t('reason', '说明'), t('reversal_of', '冲销原记录'), t('document', '凭证ID')] }] }
  let path = endpoint(resource) + r.id + '/'
  let fields: Field[] = [reason]
  let initial: Row = {}
  let method: 'post' | 'patch' = 'post'
  let readonly = false
  const routes: Record<string, string> = {
    登记收付款: 'pay',
    退款: 'refund',
    取消费用: 'cancel-expense',
    冲销付款: 'reverse',
  }
  if (routes[name]) path += routes[name] + '/'
  if (['登记收付款', '退款'].includes(name)) {
    const docs = await all('/business/documents/', { project: r.project, category: 'receipt' })
    fields = [t('amount', '金额（元）'), date(), reason, { key: 'method', label: '结算方式', type: 'select', optional: true, options: [{ value: 'bank', label: '银行转账' }, { value: 'cash', label: '现金' }, { value: 'other', label: '其他' }] }, t('account', '账户标识', true), t('reference', '银行流水号', true), { key: 'document', label: '结算凭证（先在项目附件上传）', type: 'select', optional: true, options: docs.map(d => ({ value: d.id, label: d.original_name })) }]
    initial = { amount: String(r.balance).replace(/^-/, '') }
    const statements = await all('/business/reconciliations/', { entry: r.id, status: 'confirmed' })
    const eligible = statements.filter(s => s.valid && Number(s.remaining_amount) > 0 && (s.kind === 'refund') === (name === '退款'))
    fields.push({ key: 'reconciliation', label: '已确认对账 / 预付款核准', type: 'select', optional: !(r.purchase || name === '退款'), options: eligible.map(s => ({ value: s.id, label: `${s.code} · 可用 ${s.remaining_amount}` })), hint: r.purchase || name === '退款' ? '先从款项操作生成对账单，再到业务对账页确认。预付款由项目经理或管理员核准。' : '客户到账可直接登记，不强制先对账。' })
    if (eligible.length === 1) initial.reconciliation = eligible[0]!.id
  }
  if (name === '冲销付款') fields = [date(), reason]
  return { title: name, path, fields, initial, method, readonly }
}

File: backend/apps/business/services/finance.py
from django.db.models import Sum
from django.shortcuts import get_object_or_404

from apps.core.api import Conflict
from apps.core.permissions import FINANCE

from ..models import Document, Entry, Payment, PaymentEvidence, StockMove, TimeEntry
from .common import ZERO, audit, day, fields, lookup, number, project_action, rounded, save, state, text


def paid(entry):
    if hasattr(entry, 'net_paid'):
        return entry.net_paid
    return entry.payments.aggregate(total=Sum('amount'))['total'] or ZERO


def balance(entry):
    return entry.amount - entry.credit_amount - paid(entry)


def entry_action(actor, key, operation, entry_id, data, execute):
    entry = lookup(Entry, entry_id)

    def perform(user, project):
        current = get_object_or_404(Entry.objects.select_for_update(), pk=entry.pk, project=project)
        return execute(user, current)

    return project_action(actor, key, operation, entry.project_id, {'entry': entry.pk, 'data': data}, FINANCE, perform)


def pay(actor, key, entry_id, data, *, refund=False):
    operation = 'entry.refund' if refund else 'entry.pay'

    def execute(user, entry):
        fields(data, {'amount', 'date', 'reason', 'method', 'account', 'reference', 'document', 'reconciliation'})
        # Cancelled projects must remain able to settle refunds; closed projects must reopen first.
        if entry.project.status == 'closed':
            raise Conflict('项目已结项，请先重新打开。')
        amount = number(data.get('amount'), 'amount', positive=True)
        remaining = balance(entry)
        if refund:
            if remaining >= 0 or amount > -remaining:
                raise Conflict('退款超过当前应退余额。')
            amount = -amount
        elif remaining <= 0 or amount > remaining:
            raise Conflict('付款超过当前未结余额。')
        from .reconciliation import authorize_payment

        statement = authorize_payment(entry, data, amount, refund)
        method = text(data, 'method', default='', maximum=20)
        if method not in {'', 'bank', 'cash', 'other'}:
            from rest_framework.exceptions import ValidationError

            raise ValidationError({'method': '请选择银行转账、现金或其他。'})
        document = (
            lookup(Document, data['document'], 'document', project=entry.project, category='receipt')
            if data.get('document')
            else None
        )
        payment = save(
            Payment(
                entry=entry,
                reconciliation=statement,
                amount=amount,
                date=day(data, 'date'),
                reason=text(data, 'reason'),
                method=method,
                account=text(data, 'account', default='', maximum=100),
                reference=text(data, 'reference', default='', maximum=100),
                document=document,
            ),
            user,
        )
        audit(user, operation, entry, payment=payment.pk, amount=str(amount))
        return {'id': payment.pk}

    return entry_action(actor, key, operation, entry_id, data, execute)


def attach_evidence(actor, key, payment_id, data):
    original = lookup(Payment, payment_id)

    def execute(user, entry):
        fields(data, {'document', 'reason'})
        source = get_object_or_404(Payment.objects.select_for_update(), pk=original.pk, entry=entry)
        document = lookup(Document, data.get('document'), 'document', project=entry.project, category='receipt')
        if source.document_id == document.pk or source.evidence.filter(document=document).exists():
            raise Conflict('此凭证已关联到该流水，请刷新查看。')
        evidence = save(PaymentEvidence(payment=source, document=document, reason=text(data, 'reason')), user)
        audit(user, 'payment.evidence', source, document=document.pk, reason=evidence.reason)
        return {'id': evidence.pk}

    return entry_action(
        actor, key, 'payment.evidence', original.entry_id, {'payment': original.pk, 'data': data}, execute
    )


def reverse(actor, key, payment_id, data):
    original = lookup(Payment, payment_id)

    def execute(user, entry):
        fields(data, {'date', 'reason'})
        if entry.project.status == 'closed':
            raise Conflict('项目已结项，请先重新打开。')
        source = get_object_or_404(Payment.objects.select_for_update(), pk=original.pk, entry=entry)
        if source.reversal_of_id or Payment.objects.filter(reversal_of=source).exists():
            raise Conflict('此流水不能再次冲销。')
        if source.bank_matches.filter(reversal_of__isnull=True, reversal__isnull=True).exists():
            raise Conflict('该流水已与银行记录核对，请先撤销银行匹配，再冲销原流水。')
        if paid(entry) - source.amount < 0:
            raise Conflict('冲销会使净付款为负，请先处理关联退款。')
        payment = save(
            Payment(
                entry=entry,
                amount=-source.amount,
                date=day(data, 'date'),
                reason=text(data, 'reason'),
                reversal_of=source,
            ),
            user,
        )
        return audit(user, 'payment.reverse', payment, original=source.pk)

    return entry_action(
        actor, key, 'payment.reverse', original.entry_id, {'payment': original.pk, 'data': data}, execute
    )


def expense(actor, key, data):
    def execute(user, project):
        fields(data, {'project', 'title', 'amount', 'due_date'})
        state(project, {'active', 'delivering', 'warranty'})
        entry = save(
            Entry(
                project=project,
                kind='expense',
                title=text(data, 'title', maximum=150),
                amount=number(data.get('amount'), 'amount', positive=True),
                due_date=day(data, 'due_date'),
            ),
            user,
        )
        return audit(user, 'expense.create', entry)

    return project_action(actor, key, 'expense.create', data.get('project'), data, FINANCE, execute)


def cancel_expense(actor, key, entry_id, data):
    def execute(user, entry):
        fields(data, {'reason'})
        if entry.kind != 'expense' or entry.cancelled or entry.project.status == 'closed':
            raise Conflict('此费用当前不能取消。')
        reason = text(data, 'reason')
        entry.cancelled = True
        entry.credit_amount = entry.amount
        save(entry, user)
        return audit(user, 'expense.cancel', entry, reason=reason)

    return entry_action(actor, key, 'expense.cancel', entry_id, data, execute)


def cost(project):
    return costs([project.pk])[project.pk]


def costs(project_ids):
    """Shared grouped projection for project detail, budgets and management reports."""
    from django.db.models import Q

    result = {pk: dict(materials=ZERO, purchase_return_variance=ZERO, labor=ZERO, expenses=ZERO) for pk in project_ids}
    moves = (
        StockMove.objects.filter(project_id__in=project_ids)
        .values('project_id')
        .annotate(
            materials=Sum('value', filter=Q(kind__in=['issue', 'return'])),
            returned=Sum('value', filter=Q(kind='purchase_return')),
            credit=Sum('supplier_credit', filter=Q(kind='purchase_return')),
        )
    )
    for row in moves:
        result[row['project_id']]['materials'] = -(row['materials'] or ZERO)
        result[row['project_id']]['purchase_return_variance'] = -(row['returned'] or ZERO) - (row['credit'] or ZERO)
    for row in (
        TimeEntry.objects.filter(task__project_id__in=project_ids)
        .values('task__project_id')
        .annotate(total=Sum('cost'))
    ):
        result[row['task__project_id']]['labor'] = row['total'] or ZERO
    for row in (
        Entry.objects.filter(project_id__in=project_ids, kind='expense', cancelled=False)
        .values('project_id')
        .annotate(total=Sum('amount'))
    ):
        result[row['project_id']]['expenses'] = row['total'] or ZERO
    return {
        pk: {key: str(rounded(value)) for key, value in {**values, 'total': sum(values.values(), ZERO)}.items()}
        for pk, values in result.items()
    }


def record_contract(user, project, milestones):
    """Record validated sales milestones in the caller's signing transaction."""
    for row in milestones:
        save(Entry(project=project, kind='receivable', **row), user)
